src/test/mir-opt/matches_reduce_branches.foo.PreCodegen.before.32bit.mir

```diff
@@ -0,0 +1,21 @@
+// MIR for `foo` before PreCodegen
+
+fn foo(_1: Option<()>) -> () {
+    debug bar => _1;                     // in scope 0 at $DIR/matches_reduce_branches.rs:7:8: 7:11
+    let mut _0: ();                      // return place in scope 0 at $DIR/matches_reduce_branches.rs:7:25: 7:25
+    let mut _2: isize;                   // in scope 0 at $DIR/matches_reduce_branches.rs:8:22: 8:26
+
+    bb0: {
+        _2 = discriminant(_1);           // scope 0 at $DIR/matches_reduce_branches.rs:8:22: 8:26
+        switchInt(move _2) -> [0_isize: bb2, otherwise: bb1]; // scope 0 at $DIR/matches_reduce_branches.rs:8:22: 8:26
+    }
+
+    bb1: {
```

Comment:
CC @oli-obk from https://github.com/rust-lang/rust/pull/77486#discussion_r511866322

The MIR still contains two branches in the final MIR, so the optimization in this PR still messes this up. I'm wondering if this can be considered a bug in `MatchBranchSimplification` that it does not handle the `_0 = const ()` case. 

Tangentially, is `_0 = const ()` ever needed in a function that returns unit? I think it would only make sense if `_0` is considered uninitialized at the beginning of the function. But then, it would probably be a simplification to just set `_0 = const ()` at the beginning of `bb0`.

Reply:
`_0 = const ()` is never needed, and I thought it's actually eliminated by `InstCombine` or sth. Not sure why there's still one around at this point. No ZST assignment is ever needed, you can create references to locals before they are assigned if the local is a ZST (as long as it was `StorageLive`d I think).